[`pyupgrade`] Prevent infinite loop with `I002` (`UP010`, `UP035`)

danparizher opened the pull request:
## Summary

Fixes #18729 and fixes #16802

## Test Plan

Manually verified via CLI that Ruff no longer enters an infinite loop by running:
```sh
echo 1 | ruff --isolated check - --select I002,UP010 --fix
```
with `required-imports = ["from __future__ import generator_stop"]` set in the config, confirming “All checks passed!” and no snapshots were generated.

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/unnecessary_future_import.rs:
```diff
@@ -91,6 +92,44 @@ pub(crate) fn unnecessary_future_import(checker: &Checker, stmt: &Stmt, names: &
         if alias.asname.is_some() {
             continue;
         }
+        let is_required = match stmt {
+            ruff_python_ast::Stmt::ImportFrom(ruff_python_ast::StmtImportFrom {
+                module,
+                level,
+                ..
+            }) => {
+                let name_import = NameImport::ImportFrom(MemberNameImport {
+                    module: module.as_ref().map(std::string::ToString::to_string),
+                    name: ruff_python_semantic::Alias {
+                        name: alias.name.to_string(),
+                        as_name: None,
+                    },
+                    level: *level,
+                });
+                checker
+                    .settings()
+                    .isort
+                    .required_imports
+                    .contains(&name_import)
+            }
+            ruff_python_ast::Stmt::Import(_) => {
+                let name_import = NameImport::Import(ModuleNameImport {
+                    name: ruff_python_semantic::Alias {
+                        name: alias.name.to_string(),
+                        as_name: None,
+                    },
+                });
+                checker
+                    .settings()
+                    .isort
+                    .required_imports
+                    .contains(&name_import)
+            }
+            _ => false,
+        };
```
This looks like a nice potential helper function. Refactoring it like that would both allow us to reuse some code here with an early return and share this with UP035, which I think could use the same fix (https://github.com/astral-sh/ruff/issues/16802). I'd suggest something like:

```rust
fn helper_function(/* ... */) -> bool {
  let name_import = match stmt {
              ruff_python_ast::Stmt::ImportFrom(ruff_python_ast::StmtImportFrom {
                  module,
                  level,
                  ..
              }) => {
                  NameImport::ImportFrom(MemberNameImport {
                      module: module.as_ref().map(std::string::ToString::to_string),
                      name: ruff_python_semantic::Alias {
                          name: alias.name.to_string(),
                          as_name: None,
                      },
                      level: *level,
                  })
              }
              ruff_python_ast::Stmt::Import(_) => {
                  NameImport::Import(ModuleNameImport {
                      name: ruff_python_semantic::Alias {
                          name: alias.name.to_string(),
                          as_name: None,
                      },
                  })

              }
              _ => return false,
          };
                  checker
                      .settings()
                      .isort
                      .required_imports
                      .contains(&name_import)
}
```

modulo the horrific formatting from me editing this on GitHub and with a real function name and arguments.

ntBre reviewed (commented):
Thanks! I had one suggestion to refactor the fix here, and we should also add a test. Since the fix is on the UP010 side, I think it would make sense to add a unit test in the `pyupgrade` module too. You should be able to select both rules and use the normal snapshot infrastructure otherwise.

danparizher commented:
Thanks, I extracted that as a function and also added a check into `crates\ruff_linter\src\rules\pyupgrade\rules\deprecated_import.rs`. Let me know if that's correct. For adding tests for the whole `pyupgrade` module, what would that look like? Would appreciate recommendations

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/deprecated_import.rs:
```diff
@@ -719,6 +720,12 @@ pub(crate) fn deprecated_import(checker: &Checker, import_from_stmt: &StmtImport
         return;
     }
 
+    for alias in &import_from_stmt.names {
```
Actually, this may just not be the right place to inject the check. Somewhere in here we should actually inspect the `alias` being replaced. That might be the more natural place to check `is_import_required_by_isort`.

danparizher replied:
I think the `continue` approach makes the most sense. I envision collecting skipped aliases and filtering them out during processing. Feel free to iterate on the commit how you see fit. 🙂
